sdk/metric/provider.go

```diff
@@ -0,0 +1,96 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metric // import "go.opentelemetry.io/otel/sdk/metric"
+
+import (
+	"context"
+
+	"go.opentelemetry.io/otel/metric"
+)
+
+// MeterProvider handles the creation and coordination of Meters. All Meters
+// created by a MeterProvider will be associated with the same Resource, have
+// the same Views applied to them, and have their produced metric telemetry
+// passed to the configured Readers.
+type MeterProvider struct {
+	// TODO (#2820): implement.
+}
+
+// Compile-time check MeterProvider implements metric.MeterProvider.
+var _ metric.MeterProvider = (*MeterProvider)(nil)
+
+// NewMeterProvider returns a new and configured MeterProvider.
+//
+// By default, the returned MeterProvider is configured with the default
+// Resource and no Readers.
+func NewMeterProvider(options ...Option) *MeterProvider {
+	// TODO (#2820): implement.
+	return &MeterProvider{}
+}
+
+// Meter returns a Meter with the given name and configured with options.
+//
+// The name should be the name of the instrumentation library creating
+// telemetry. This name may be the same as the instrumented code only if that
+// code provides built-in instrumentation.
+//
+// If name is empty, the default (go.opentelemetry.io/otel/sdk/meter) will be
+// used.
+//
+// Calls to the Meter method after Shutdown has been called will return Meters
+// that perform no operations.
+//
+// This method is safe to call concurrently.
+func (mp *MeterProvider) Meter(name string, options ...metric.MeterOption) metric.Meter {
+	// TODO (#2821): ensure this is concurrent safe.
+	// TODO: test this is concurrent safe.
+	// TODO (#2821): register and track the created Meter.
+	return &meter{}
+}
+
+// Flush flushes all pending telemetry.
+//
+// This method honors the deadline or cancellation of ctx. An appropriate
+// error will be returned in these situations. There is no guaranteed that all
+// telemetry be flushed or all resources have been released in these
+// situations.
+//
+// This method is safe to call concurrently.
+func (mp *MeterProvider) Flush(ctx context.Context) error {
+	// TODO (#2820): implement.
+	// TODO: test this is concurrent safe.
+	return nil
+}
```

Comment:
Could we punt on a Flush method?  

I think it may be added as part of the reader discussion, but absent that I don't think this will do anything.  If we move forward with the idea that Reader encapsulates MeterProviders then Flush will always be a No Op, because the MeterProvider wouldn't have access to the Reader or its exporter's flush.

Reply:
The SDK spec has a ForceFlush method. I don't believe we should be discussing "Reader encapsulates MeterProvider", at least I'm not sure why it's being proposed. When the user flushes or shuts down the SDK, all of the reader/exporters will be flushed or shut down.

Reply:
> Could we punt on a Flush method?
> 
> I think it may be added as part of the reader discussion, but absent that I don't think this will do anything. If we move forward with the idea that Reader encapsulates MeterProviders then Flush will always be a No Op, because the MeterProvider wouldn't have access to the Reader or its exporter's flush.

The flush method is a [required method of the `MeterProvider`](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#forceflush). I don't think we can punt on it even if we wanted to go in a different control flow order direction (something I doubt we'll do based on @jmacd's comments) and still remain compliant with the specification.